htdocs/core/lib/functions.lib.php

```diff
@@ -1649,6 +1649,55 @@ function dol_buildpath($path, $type = 0, $returnemptyifnotfound = 0)
 	return $res;
 }
 
+/**
+ * Return path of url.
+ *
+ * @param	string							$path		Relative path to file
+ * @param	array<string,int|float|string>	$params     params for the http query
+ * @param	bool							$addtoken	does we need to add token
+ * @param	bool 	 	 					$addurlroot Add DOL_URL_ROOT
+ * @return string										path
+ */
+function dolBuildUrl($path, $params = [], $addtoken = false, $addurlroot = false)
+{
+	global $conf, $db, $hookmanager;
+
+	if (!is_object($hookmanager)) {
+		include_once DOL_DOCUMENT_ROOT . '/core/class/hookmanager.class.php';
+		$hookmanager = new HookManager($db);
+	}
+	if ($addurlroot) {
+		$path .= DOL_URL_ROOT . $path;
+	}
+	if ((!isset($params['mainmenu']) || empty($params['mainmenu'])) && GETPOSTISSET('mainmenu')) {
+		$params = array_merge($params, ['mainmenu' => (GETPOST('mainmenu', 'restricthtml'))]);
+	}
+	if ((!isset($params['leftmenu'])/*  || empty($params['leftmenu']) */) && GETPOSTISSET('leftmenu')) { // do not fill leftmenu if we have leftmenu=
+		$params = array_merge($params, ['leftmenu' => (GETPOST('leftmenu', 'restricthtml'))]);
+	}
+	// TODO check this with multicompany
+	// if (isModEnabled('multicompany') && !isset($params['entity'])) {
+	// 	$params = array_merge($params, ['entity' => $conf->entity]);
+	// }
+	$parameters = [
+		'path' => &$path,
+		'params' => &$params,
+		'addtoken' => &$addtoken,
+	];
+	$hookmanager->executeHooks('buildurl', $parameters);
+	$url = dol_buildpath($path, 1);
```

Comment:
This has 2 very bad side effects:
- First, the path may not be the one we want because dol_buildpath is searching for alternative path into modules when we want to be sure to use a core file when doing a DOL_URL_ROOT.'/thepath'
This is due to the part of code inside dol_buildpath inside loop 
`foreach ($conf->file->dol_document_root as $key => $dirroot) {`
- The second pb is that we do a file_exist() at every call of dolBuildUrl on a file that often does not eists (so with no cache managed by dolibarr). This is consumming a very high IO.

We can solve both case by replacing this line
`$url = dol_buildpath($path, 1);`
with
```
$tmppath = preg_replace('/^\//', '', $path);
$url = DOL_URL_ROOT.'/'.$tmppath;
```

Reply:
to avoid multiple parameters, I keep the original DOL_URL_ROOT or dol_builpath if present